text/0047-storage.md

```diff
@@ -0,0 +1,217 @@
+- Consider only RocksDB in the begining for ease of implementation and compatible with the current [IRI](https://github.com/iotaledger/iri) design
+- File hierarchy
+
+```
+    +-- access
+    |   +-- transaction.rs
+    |   +-- milestone.rs
+    |   +-- state_delta.rs
+    |   +-- tangle.rs
+    |   +-- snapshot.rs
+    |   +-- address.rs
+    |   +-- ..
+    +-- storage
+    |   +-- config.rs
+    |   +-- mod.rs
+    |   +-- rocksdb.rs
+    |   +-- ..
+```
+
+- Current Structures
+
+```rust
+/// The Storage struct will be created automatically by the assignment in feature
+/// Here we use rocks_db as an example for the storage structure
+
+#[cfg(feature = "rocks_db")]
+pub struct Storage {
+    inner: ::rocksdb::DB  // The storage backend
+}
+
+// From https://github.com/iotaledger/bee-rfcs/pull/30
+#[derive(Copy, Clone)]
+pub struct Milestone {
+    pub(crate) hash: Hash,
+    pub(crate) index: MilestoneIndex,
+}
+
+/// From https://github.com/iotaledger/bee-rfcs/pull/20
+#[derive(Clone, Debug, PartialEq, Hash)]
+pub struct Address(pub(crate) TritBuf<T1B1Buf>);
+
+/// From https://github.com/iotaledger/bee-rfcs/pull/20
+#[derive(PartialEq, Clone, Debug)]
+pub struct BundledTransaction {
+    pub(crate) payload: Payload,
+    pub(crate) address: Address,
+    pub(crate) value: Value,
+    pub(crate) obsolete_tag: Tag,
+    pub(crate) timestamp: Timestamp,
+    pub(crate) index: Index,
+    pub(crate) last_index: Index,
+    pub(crate) bundle: Hash,
+    pub(crate) trunk: Hash,
+    pub(crate) branch: Hash,
+    pub(crate) tag: Tag,
+    pub(crate) attachment_ts: Timestamp,
+    pub(crate) attachment_lbts: Timestamp,
+    pub(crate) attachment_ubts: Timestamp,
+    pub(crate) nonce: Nonce,
+}
+```
+
+- Backend trait and the associated methods
+    - start method
+    - shutdown method
+
+```rust
+#[async_trait]
+/// Trait to be implemented on storage backend,
+/// which determine how to start and shutdown the storage
+pub trait Backend {
+    /// start method should impl how to start and initialize the corrsponding database
+    /// It takes config_path which define the database options, and returns Result<Self, Box<dyn Error>>
+    async fn start(config_path: String) -> Result<Self, Box<dyn Error>> where Self: Sized;
+    /// shutdown method should impl how to shutdown the corrsponding database
+    /// It takes the ownership of self, and returns ()
+    async fn shutdown(self) -> Result<(), Box<dyn Error>>;
+}
+#[cfg(feature = "rocks_db")]
+#[async_trait]
+impl Backend for Storage {
+    /// It starts RocksDB instance and then initialize the required column familes
+    async fn start(config_path: String) -> Result<Self, Box<dyn Error>> {
+        let config_as_string = fs::read_to_string(config_path)?;
+        let config: config::Config = toml::from_str(&config_as_string)?;
```

Comment:
Where is this `Config` type from? The `config` crate? If so, it might make more sense to have the config loading abstracted such that storage API users can customise it.

Reply:
Yes it's already abstracted using feature, you can check that in storage branch.